hail/src/main/scala/is/hail/io/fs/AzureStorageFS.scala

```diff
@@ -0,0 +1,388 @@
+
+      override def close(): Unit = {
+        if (!closed) {
+          flush()
+          closed = true
+        }
+      }
+    }
+
+    new WrappedPositionedDataOutputStream(os)
+  }
+
+  def delete(filename: String, recursive: Boolean): Unit = {
+    if (this.exists(filename)) {
+      val (account, container, path) = getAccountContainerPath(filename)
+      val blobClient: BlobClient = this.getBlobClient(filename)
+
+      val fileStatus = this.fileStatus(filename)
+
+      if (recursive) {
+        val blobContainerClient = this.getContainerClient(filename)
+
+        val directoryContents = blobContainerClient.listBlobsByHierarchy(dropTrailingSlash(path) + "/")
+        var seenBlobItems: Set[String] = Set()
+        val iter = directoryContents.iterator()
+
+        def deleteRecursively(): Unit = {
+          while (iter.hasNext) {
+            val blobItem: BlobItem = iter.next
+            if (seenBlobItems.contains(blobItem.getName)) return
+            seenBlobItems += blobItem.getName
+            val blobFileName = s"hail-az://$account/$container/${blobItem.getName}"
+            // recursively delete all files/subfolders
+            this.delete(blobFileName, recursive)
+          }
+        }
+        deleteRecursively()
+
+        if (fileStatus.isFile && blobClient.exists()) blobClient.delete()
+      }
+      else {
+          if (fileStatus.isDirectory) {
+            throw new IllegalArgumentException("Cannot non-recursively delete a directory")
+          }
+          if (blobClient.exists()) blobClient.delete()
+      }
+    }
+  }
+
+  def listStatus(filename: String): Array[FileStatus] = {
+    val (account, container, path) = getAccountContainerPath(filename)
+    val blobContainerClient: BlobContainerClient = getContainerClient(filename)
+
+    val statusList: mutable.MutableList[FileStatus] = mutable.MutableList()
+    var blobNameSet: Set[String] = Set()
+
+    var seenBlobItems: Set[String] = Set()
+
+    val prefix = dropTrailingSlash(path) + "/"
+    val directoryContents = blobContainerClient.listBlobsByHierarchy(prefix)
+    val iter = directoryContents.iterator()
+
+    def populateStatList(): Unit = {
+      while (iter.hasNext) {
+        val blobItem: BlobItem = iter.next
+
+        if (seenBlobItems.contains(blobItem.getName)) return
+        seenBlobItems += blobItem.getName
+
+        // find the first slash after the prefix
+        val indexOfFirstSlashAfterPrefix: Int = blobItem.getName.indexOf("/", prefix.length)
+        var blobName = blobItem.getName
+        // if there is a slash after the prefix, then this blob is under a subdirectory
+        if (indexOfFirstSlashAfterPrefix != -1) {
+          blobName = blobItem.getName.substring(0, indexOfFirstSlashAfterPrefix)
+        }
+
+        if (!blobNameSet.contains(blobName)) {
+          blobNameSet = blobNameSet + blobName
+          val blobFileName = s"hail-az://$account/$container/$blobName"
+          statusList += fileStatus(blobFileName)
```

Comment:
We already have the `BlobItem` at this point, so we should take advantage of the information it holds and assemble a `BlobStorageFileStatus` directly from the `BlobProperties`.

Reply:
See my below comment about making an `apply` method for `AzureStorageFileStatus`

Reply:
This is not as easy as I would like it to be since AzureStorage will return a 404 error if you call getProperties on a path that corresponds to a directory (and not directly to a blob)